### PR TITLE
[T2-Chassis] - New tests for Reliable-TSA feature

### PR DESCRIPTION
**Pull request by sanjair-git**
### Description of PR

Summary:
Fixes # (issue)

- This PR has tests to verify the changes introduced as part of #https://github.com/sonic-net/sonic-buildimage/pull/19217. The original code PR for the feature is #https://github.com/sonic-net/sonic-buildimage/pull/18928
- HLD link for the feature: https://github.com/skeesara-nokia/SONiC/blob/master/doc/voq/Reliable_TSA.md
- Basically, this PR has 20 test cases to verify the feature functionality and also few modifications on the tests introduced as part of PR #12781 

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [x] 202405

#### What is the motivation for this PR?

- To verify the feature functionality and the code changes introduced as part of the PR #https://github.com/sonic-net/sonic-buildimage/pull/18928

#### How did you do it?

-  Added 20 new test cases to verify feature in three sets. 
- Attaching test cases' description document here for further reference. 
- [Reliable-TSA-Cases.xlsx](https://github.com/user-attachments/files/15827682/Reliable-TSA-Cases.xlsx)

1. First set:  TSA/TSB config event on the supervisor or line card. (8 cases)
     _test_sup_tsa_act_when_sup_duts_on_tsb_initially
     test_sup_tsa_act_when_sup_on_tsb_duts_on_tsa_initially
     test_sup_tsb_act_when_sup_on_tsa_duts_on_tsb_initially
     test_sup_tsb_act_when_sup_and_duts_on_tsa_initially
     test_dut_tsa_act_when_sup_duts_on_tsb_initially
     test_dut_tsa_act_when_sup_on_tsa_duts_on_tsb_initially
     test_dut_tsb_act_when_sup_on_tsb_duts_on_tsa_initially
     test_dut_tsb_act_when_sup_and_duts_on_tsa_initially_

2. Second set: Config event followed by certain action on the supervisor or line card. (8 cases)
     _test_sup_tsa_act_with_sup_reboot
     test_sup_tsa_act_when_duts_on_tsa_with_sup_config_reload
     test_dut_tsa_act_with_reboot_when_sup_dut_on_tsb_init
     test_dut_tsa_with_conf_reload_when_sup_on_tsa_dut_on_tsb_init
     test_sup_tsb_followed_by_dut_bgp_restart_when_sup_on_tsa_duts_on_tsb
     test_sup_tsb_followed_by_dut_bgp_restart_when_sup_and_duts_on_tsa
     test_dut_tsb_followed_by_dut_bgp_restart_when_sup_on_tsb_duts_on_tsa
     test_dut_tsb_followed_by_dut_bgp_restart_when_sup_and_duts_on_tsa_
     
3. Miscellaneous tests: Use cases with '_startup-tsa-tsb_' service etc. (4 cases)
     _test_sup_tsa_when_startup_tsa_tsb_service_running
     test_sup_tsb_when_startup_tsa_tsb_service_running
     test_user_init_tsa_on_dut_followed_by_sup_tsa
     test_user_init_tsa_on_dut_followed_by_sup_tsb_

#### How did you verify/test it?

- Ran all the above-mentioned test cases on a T2 chassis and made sure tests passed with expected behavior.

### Documentation

**Reliable TSA test result**:
image

**Startup-tsa-tsb service test result:**
image

**Note:** PR #12781 is pending merge, and this PR is dependent on that.

### CODE REVIEW

File: tests/bgp/test_reliable_tsa.py
Comment (tjchadaga) [inline tests/bgp/test_reliable_tsa.py:311]: TSA persistence support on the linecard exists on all images that supports T2 and since this is a T2-only test, this check is not needed. 

Should we be checking chassis-app-db for tsa_enabled field instead and skip if reliable tsa is not in the image?

File: tests/bgp/test_reliable_tsa.py
Comment (sanjair-git) [reply tests/bgp/test_reliable_tsa.py]: Yes right. Modified the code to re-use from test_startup_tsa_tsb_service.py file.

**Comment by sanjair-git** [comment]
This test PR is dependent on sonic-buildimage PR #(https://github.com/sonic-net/sonic-buildimage/pull/19539)

**Comment by abdosi** [comment]
@tjchadaga: Are we good to merge this ?

**Comment by sanjair-git** [comment]
**Latest test results for reference:**

image

**Comment by yejianquan** [comment]
@bingwang-ms (202405 release manager for viz)
Since this's a new test, reliable TSA /TSB that needed in chassis test and we want to cherry-pick to 202405
I have confirmed, most of them are with the test mark that runs only on T2,
few of them are with the test mark T1,T2, but the change only influences the topology with T2.

So it wouldn't affect T1 tests and cause regression

**Comment by yejianquan** [comment]
> @bingwang-ms (202405 release manager for viz) Since this's a new test, reliable TSA /TSB that needed in chassis test and we want to cherry-pick to 202405 I have confirmed, most of them are with the test mark that runs only on T2, few of them are with the test mark T1,T2, but the change only influences the topology with T2.
> 
> So it wouldn't affect T1 tests and cause regression

And test_traffic_shift is covered by the PR on kvm t1-lag testbeds, it passed, test: https://elastictest.org/scheduler/testplan/66eda86ec4757bff3198c203?searchTestCase=traffic&testcase=bgp%2Ftest_traffic_shift.py&type=console

File: tests/bgp/test_startup_tsa_tsb_service.py
Comment (Javier-Tan) [inline tests/bgp/test_startup_tsa_tsb_service.py:255]: Would it be good to indicate what caused the DUT to not be normal state?

**Review by Javier-Tan — COMMENTED** [review]
General comments, generally LGTM 👍

File: tests/bgp/bgp_helpers.py
Comment (sanjair-git) [reply tests/bgp/bgp_helpers.py]: Corrected in the latest commit. Please check

File: tests/bgp/test_reliable_tsa.py
Comment (sanjair-git) [reply tests/bgp/test_reliable_tsa.py]: Yea true. Removed from this test file.

File: tests/bgp/test_startup_tsa_tsb_service.py
Comment (sanjair-git) [reply tests/bgp/test_startup_tsa_tsb_service.py:257]: Added a clear logger statement indicating details.

**Review by Javier-Tan — APPROVED** [review]
LGTM, if you can please list your general updates/QoL test changes that affect my PR in the same files #14855

**Comment by sanjair-git** [comment]
> @sanjair-git, can please resolve conflicts

@arlakshm, I have resolved the conflicts and taken care of the new test added as part of #14855. Please help in reapproving and merge this PR. Thanks

**Review by abdosi — APPROVED** [review]

**Review by tjchadaga — APPROVED** [review]

**Comment by vperumal** [comment]
Hi @abdosi @yejianquan - Can you help getting it merged to 202405 - it has been 4 days since it has been merged to master ?

**Pull request merged**
